File: Homework6/phonebook/impl/PhoneBookImpl.java
package Homework6.phonebook.impl;

import Homework6.model.Contact;
import Homework6.phonebook.PhoneBook;

import java.util.List;

public class PhoneBookImpl implements PhoneBook {

    private List<Contact> contacts;

    public PhoneBookImpl(List<Contact> contacts) {
        this.contacts = contacts;
    }


    public String addContact(Contact contact) {
        if (!isExist(contact)) {
            contacts.add(contact);
            return "Ok";
        }
        return "Contact exist";
    }

    public String deleteContact(int index) {
        Contact contact = findContact(index);
        if (!isExist(contact)) {
            return "Contact not existing";
        }
        contacts.remove(contact);
        return "Contact deleted";
    }

    public Contact updateContact(int index, Contact contact) {
        if (!isExist(contact)) {
            addContact(contact);
            return contact;
        }

        contacts.set(index, contact);
        return contact;
    }

    public Contact findContact(int index) {
        if (index < 0 || index > contacts.size()) {
            throw new IndexOutOfBoundsException();
        }
        return contacts.get(index);

    }

    public void getInfoWithContacts() {
        contacts.forEach(System.out::println);
    }

    private boolean isExist(Contact contact) {
        return contacts.contains(contact);
    }
}

File: Homework7/phonebook/impl/PhoneBookImpl.kt
package Homework7.phonebook.impl

import Homework7.model.Contact
import Homework7.phonebook.PhoneBook
import java.util.function.Consumer

class PhoneBookImpl(private val contacts: MutableList<Contact>) : PhoneBook {
    override fun addContact(contact: Contact?): String {
        if (!isExist(contact)) {
            if (contact != null) {
                contacts.add(contact)
            }
            return "Ok"
        }
        return "Contact exist"
    }

    override fun deleteContact(index: Int): String {
        val contact = findContact(index)
        if (!isExist(contact)) {
            return "Contact not existing"
        }
        contacts!!.remove(contact)
        return "Contact deleted"
    }

    override fun updateContact(index: Int, contact: Contact?): Contact? {
        if (!isExist(contact)) {
            addContact(contact)
            return contact
        }
        if (contact != null) {
            contacts[index] = contact
        }
        return contact
    }

    override fun findContact(index: Int): Contact {
        if (index < 0 || index > contacts.size) {
            throw IndexOutOfBoundsException()
        }
        return contacts[index]
    }

    override val infoWithContacts: Unit
        get() {
            contacts.forEach(Consumer { x: Contact? -> println(x) })
        }

    private fun isExist(contact: Contact?): Boolean {
        return contacts.contains(contact)
    }
}